### PR TITLE
ENH - More efficient ``B.dot`` and ``B.T.dot`` in Cox datafit

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
In [Cox datafit](https://contrib.scikit-learn.org/skglm/tutorials/cox_datafit.html#maths-cox-datafit), It is possible to perform the operations evolving the matrix `B` in linear time and without having to form explicitly the matrix.

Up to a sorting permutation of ``tm``, the matrix B has a block triangular structure involving matrices with coordinates equal to 1.
Hence applying `B` to a vector (slightly) resembles performing a cumulative sum on the coordinates of `v`.

### Advantage

- reduce the memory footprint of Cox datafit from O(n²) --> O(n) as we no longer store ``B``
- Evaluate gradient and Hessian in O(n) instead of O(n²) 

### Benchmarks

A link to[ benchmark results](https://badr-moufad.github.io/share-benchmarks/cox_bench/benchmark_cox_benchopt_run_2023-06-14_16h29m06.html) to showcase the timing improvements compared to the ``main`` branch.

---------

Riding "la trottinette :kick_scooter: à la @agramfort"

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
Need to merge #167 beforehand

**Comment by Badr-MOUFAD** [comment]
The test fails because the gradient and Hessian disagree with `scipy check_grad` at `atol=1e-3`.
I'm not that familiar with truncation and roundoff errors to blame numerical errors for that.

However, I pushed a `debug_script` that checks independently the matrix-vector operations. It reveals that both operations are the same (difference of 1e-11).

@mathurinm, any clues?

File: skglm/datafits/single_task.py [hunk 1]
```diff
@@ -561,19 +561,24 @@ class Cox(BaseDatafit):
 
     Attributes
     ----------
-    B : array-like, shape (n_samples, n_samples)
-        Matrix where every ``(i, j)`` entry (row, column) equals ``1``
-        if ``tm[j] >= tm[i]`` and ``0`` otherwise. This matrix is initialized
-        using the ``.initialize`` method.
+    T_indices : array-like, shape (n_samples,)
+        Indices of observations with the same occurrence times stacked horizontally as
+        ``[group_1, group_2, ...]`` in an ascending order. It is initialized
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:566]:
```suggestion
        ``[group_1, group_2, ...]`` in ascending order. It is initialized
```

File: skglm/datafits/single_task.py [hunk 2]
```diff
@@ -561,19 +561,24 @@ class Cox(BaseDatafit):
...
+        with ``.initialize`` (or ``initialize_sparse`` for sparse ``X``) method.
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:567]: *the* XX method ?

File: skglm/datafits/single_task.py [hunk 3]
```diff
@@ -561,19 +561,24 @@ class Cox(BaseDatafit):
...
+    T_indptr : array-like, (np.unique(tm) + 1,)
+        Array where two consecutive elements delimit a group of
+        observations having the same occurrence times.
 
-    H_indptr : array-like, (np.unique(tm) + 1,)
-        Array where two consecutive elements delimits a group of observations
-        having the same occurrence times.
+    H_indices : array-like, shape (n_samples,)
+        Indices of uncensored observations with the same occurrence times stacked
+        horizontally as ``[group_1, group_2, ...]`` in an ascending order.
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:575]: ditto

File: skglm/datafits/single_task.py [hunk 4]
```diff
@@ -654,38 +658,53 @@ def initialize(self, X, y):
         """Initialize the datafit attributes."""
         tm, s = y
 
-        tm_as_col = tm.reshape((-1, 1))
-        self.B = (tm >= tm_as_col).astype(X.dtype)
+        self.T_indices = np.argsort(tm)
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:661]: nice!

File: skglm/datafits/single_task.py [hunk 5]
```diff
@@ -712,3 +732,21 @@ def _AT_dot_vec(self, vec):
...
+
+    def _get_indptr(self, vals, indices):
+        # given `indices = argsort(vals)`
+        # build and array `indptr` where two consecutive elements
+        # delimit indices with the same val
+        n_indices = indices.shape[0]
+
+        indptr = [0]
+        count = 1
+        for i in range(1, n_indices):
```
Comment (mathurinm) [inline skglm/datafits/single_task.py:744]: is it not more natural to have range(n_indices - 1) and then use i and i + 1 L745?

**Review by mathurinm — APPROVED** [review]

**Pull request merged**